File: node/day1part2.js
#!/usr/bin/env node
/*
 * A Javascript / NodeJS implementation of AOC 2021 Day 1 Part 2
 * with sliding windows and everything.
 *
 * Mainly to prove I can.
 */

fs = require('fs');

//const inputfile = "data/day1test.txt";
const inputfile = "data/day1part1.txt";

const readings = fs.readFileSync(inputfile,'UTF-8').split(/\r?\n/);

const windows = [];
const winsize = 3;

readings.forEach(reading => {
  const ping = Number(reading);
  //dummy out if NaN/zero
  if(ping==0) { return }
  windows.push(ping);
  ptr = windows.length;
  if(ptr < winsize) {
    i=0;
    while(i<(ptr-1)) {
      windows[i] = windows[i] + ping;
      i++;
    }
  } else {
    m=ptr-winsize;
    while(m<(ptr-1)) {
      windows[m] = windows[m] + ping;
      m++;
    }
  }
});
//Last winsize-1 entries are incomplete and should be removed
const compWindows = windows.slice(0,(windows.length - winsize + 1))

//And for the comparison we just need to iterate over the elements comparing
//to last:
increases = 0;
for(var i = 0; i<=compWindows.length; i++) {
  if(i==0) { continue } //can't do anything with first entry
  if(compWindows[i] > compWindows[i-1]) { increases++; }
}
console.log("Over " + compWindows.length + " measurement windows, found " + increases + " reading increases.")

File: python/day1part2.py
#!/usr/bin/python3
# Python solution for AOX 2021 Day 1, Part 2
# Given an input "sonar depth readings", use a 3-measurement sliding window
# to assess increases, reporting the number of increases.

#inputfile = "data/day1test.txt"
inputfile = "data/day1part1.txt"

lc=0
windows=[]
winsize=3
#This is the reading loop
with open(inputfile,'r') as sonarreadings:
    for reading in sonarreadings:
        lc+=1
        ping=int(reading)
        windows.append(ping)
        ptr = len(windows) -1
        if ptr < winsize :
            for i in range(0,ptr) :
                windows[i] += ping
        else :
            for i in range(1,winsize) :
                x = ptr-i
                windows[x] += ping
#the last winsize-1 readings are bogus, drop them
for i in range(1,winsize) :
    windows.pop()

#The answer is just the same as part 1, but now we've got a list it's easier to compute
increases=0
for i in range(len(windows)) :
    if i == 0 :
        continue
    if windows[i] > windows[i-1] :
        increases+=1

print(f"Over {len(windows)} measurement windows, found {increases} reading increases")
